hail/python/hailtop/batch/job.py

```diff
@@ -909,6 +912,10 @@ def wrapped(*args, **kwargs):
             job_path = os.path.dirname(result._get_path(remote_tmpdir))
             code_path = f'{job_path}/code{i}.p'
 
+            if self._batch._gcs is None:
+                self._batch._gcs = GCS(blocking_pool=concurrent.futures.ThreadPoolExecutor(),
+                                       project=self._batch._project)
```

Comment:
This feels like unnecessary deep entanglement between these classes.

Let's make `_gcs` a lazy field:

```python3
    self.__gcs: Optional[GCS] = None

    @property
    def _gcs(self):
        if self.__gcs is None:
            self.__gcs = GCS(...)
        return self.__gcs
```